api/handlers/videostream.go

```diff
@@ -35,54 +35,340 @@ LICENSE
+	if err != nil {
+		return api.InvalidRequestURL(ctx)
+	}
+
+	// Fetch data from the datastore.
+	store := ds_client.Get()
+	key := store.IDKey("VideoStream", id)
+	var videoStream model.VideoStream
+	if store.Get(context.Background(), key, &videoStream) != nil {
+		return api.DatastoreReadFailure(ctx)
+	}
+
+	// Format result.
+	result := FromVideoStream(&videoStream, int(id), format)
 
-	// id, _ := ctx.ParamsInt("id", 1)
-	return ctx.JSON("TODO")
+	return ctx.JSON(result)
 }
 
 // GetVideoStreams gets a list of video streams, filtering by timespan, capture source if specified.
 func GetVideoStreams(ctx *fiber.Ctx) error {
+	// Parse URL.
 	qry := new(GetVideoStreamsQuery)
 	qry.SetLimit()
 
 	if err := ctx.QueryParser(qry); err != nil {
 		return api.InvalidRequestURL(ctx)
 	}
 
-	// Placeholder code: returns an empty result.
-	// TODO: implement fetching from datastore.
-	result := api.Result[VideoStreamResult]{
-		Results: []VideoStreamResult{},
+	format := new(api.Format)
+	if err := ctx.QueryParser(format); err != nil {
+		return api.InvalidRequestURL(ctx)
+	}
+
+	// Fetch data from the datastore.
+	store := ds_client.Get()
+	query := store.NewQuery("VideoStream", false)
+
+	// Filter by timespan.
+	if qry.TimeSpan != nil {
+		if qry.TimeSpan.Start.After(qry.TimeSpan.End) { // Validate timespan
+			return api.InvalidRequestURL(ctx)
+		}
+
+		query.FilterField("StartTime", "<=", qry.TimeSpan.End)
+		query.FilterField("EndTime", ">=", qry.TimeSpan.Start)
+	}
+
+	// Filter by capture source.
+	if qry.CaptureSource != nil {
+		query.FilterField("CaptureSource", "=", *qry.CaptureSource)
+	}
+
+	query.Limit(qry.Limit)
+	query.Offset(qry.Offset)
+
+	var videoStreams []model.VideoStream
+	keys, err := store.GetAll(context.Background(), query, &videoStreams)
+	if err != nil {
+		return api.DatastoreReadFailure(ctx)
+	}
+
+	// Format results.
+	results := make([]VideoStreamResult, len(videoStreams))
+	for i := range videoStreams {
+		results[i] = FromVideoStream(&videoStreams[i], int(keys[i].ID), format)
+	}
+
+	return ctx.JSON(api.Result[VideoStreamResult]{
+		Results: results,
 		Offset:  qry.Offset,
 		Limit:   qry.Limit,
-		Total:   0,
+		Total:   len(results),
+	})
+}
+
+// putVideoStream puts a video stream in the datastore, checking if the capture source exists.
+func putVideoStream(ctx *fiber.Ctx, vs model.VideoStream) (int, error) {
+	// Verify CaptureSource exists.
+	store := ds_client.Get()
+	key := store.IDKey("CaptureSource", vs.CaptureSource)
+	var captureSource model.CaptureSource
+	if store.Get(context.Background(), key, &captureSource) != nil {
+		return 0, api.DatastoreReadFailure(ctx)
 	}
-	return ctx.JSON(result)
+
+	// Get a unique ID for the new video stream.
+	key = store.IncompleteKey("VideoStream")
+
+	var err error
+	key, err = store.Put(context.Background(), key, &vs)
+	if err != nil {
+		print(err.Error())
+		return 0, api.DatastoreWriteFailure(ctx)
+	}
+
+	// Return ID of created video stream.
+	return int(key.ID), nil
```

Comment:
Why not just have this function return the type of key.ID ?

Reply:
Agreed, changed type to int64